src/raftstore/store/worker/read.rs

```diff
@@ -364,51 +397,158 @@ impl<C: Sender<StoreMsg>> LocalReader<C> {
+            None
+        }
+    }
+
+    fn handle_tasks(&mut self, tasks: &mut Vec<Task>) -> Option<Instant> {
+        let mut sent = None;
+        for task in tasks {
+            match task.action.take() {
+                Some(Action::Register(delegate)) => {
+                    debug!(
+                        "{} register ReadDelegate for {:?}",
+                        delegate.tag, delegate.peer_id
+                    );
+                    self.delegates.insert(delegate.region.get_id(), delegate);
+                }
+                Some(Action::Read(StoreMsg::RaftCmd {
+                    send_time,
+                    request,
+                    callback,
+                })) => {
+                    if let Some(cmd) = self.propose_raft_command(request, callback, send_time) {
+                        task.action = Some(Action::Read(cmd));
+                    }
+                    if sent.is_none() {
+                        sent = Some(send_time);
+                    }
+                }
+                Some(Action::Read(StoreMsg::BatchRaftSnapCmds {
+                    send_time,
+                    batch,
+                    on_finished,
+                })) => {
+                    if let Some(cmd) =
+                        self.propose_batch_raft_snapshot_command(batch, on_finished, send_time)
+                    {
+                        task.action = Some(Action::Read(cmd));
+                    }
+                    if sent.is_none() {
+                        sent = Some(send_time);
+                    }
+                }
+                Some(Action::Update((region_id, progress))) => {
+                    if let Some(delegate) = self.delegates.get_mut(&region_id) {
+                        delegate.update(progress);
+                    } else {
+                        panic!(
+                            "unregistered ReadDelegate, region_id: {}, {:?}",
+                            region_id, progress
+                        );
                     }
-                    ret.push(Some(ReadResponse {
-                        response,
-                        snapshot: None,
-                    }));
+                }
+                Some(Action::Destroy(region_id)) => {
+                    self.delegates.remove(&region_id);
+                }
+                _ => {
+                    unimplemented!("unsupported Msg {}", task);
                 }
             }
         }
+        sent
+    }
+
+    fn handle_ready_tasks(&mut self, tasks: &mut Vec<Task>) {
+        // Dont check region epoch.
+        let mut executor = ReadExecutor::new(self.kv_engine.clone(), false);
+        // TODO: should we issue an fence to guarantee we take snapshot first
+        //       and then read current timespce?
+        executor.set_snapshot_time(monotonic_raw_now());
 
-        on_finished.invoke_batch_read(ret);
+        for mut task in tasks.drain(..) {
+            if task.action.is_none() {
```

Comment:
I think `handle_tasks` can return if all tasks' actions are none.

Reply:
Reflected in https://github.com/tikv/tikv/pull/3443/commits/c456b1faa6d50d84953d7658abfe829664000738 .